C++/Toolbox/Option/Option.h

```diff
@@ -0,0 +1,127 @@
+#ifndef _OPTION_H_
+#define _OPTION_H_
+
+#include <functional>
+#include <memory>
+#include <type_traits>
+
+namespace Toolbox {
+namespace Option {
+
+  /* The fundamental class describing an Option. Users only collect the Opt<T> as a result and do
+     do not instantiate this class directly. The static methods Some, None, and Option have 
+     standalone function aliases for convenience.
+
+     The underlying data structure is a std::unique_ptr. An Opt<T> is move only. 
+
+     The fundamental operations are forEach, map, and filter. `forEach` simply allows access to the
+     underlying data if it is Some. `map` and `filter` allow access and modification of the Opt<T>.
+  */ 
+  template<typename T>
+  class Opt
+  {
+  public:
+    // forward arguments to construct a Some<T>
+    template<typename... Args>
+    static Opt Some(Args... args)
+    {
+      return Option::Opt<T>(std::move(std::make_unique<T>(std::forward<Args>(args)...)));
+    }
+
+    // Create a None<T>
+    static Opt None()
+    {
+      return Option::Opt<T>();
+    }
+
+    // Create an Option from a pointer
+    // None<T> if is nullptr
+    // Else Some<T>
+    static Opt Option(T* t)
+    {
+      return Option::Opt<T>(std::move(std::unique_ptr<T>(t)));
+    }
+
+  private:
+    // Hidden constructors - one can't instantiate an Opt directly, only the class can
+    Opt(void) = default;
+    Opt(std::unique_ptr<T> p) : m_pData(std::move(p)) {};
+
+  public:
+    // No copy constructor available, explicitly delete so that it so intent is clear
+    template<typename T>
+    Opt(const Opt&) = delete;
+
+  public:
+    // operator bool for quick check
+    operator bool() const {
+      return (m_pData != nullptr);
+    };
+
+  public:
+    // forEach - read the value if it exists
+    void forEach(std::function<void(const T&)> forEachFunction) {
+      if (m_pData)
+        forEachFunction(*m_pData.get());
+    };
+
+    // map - concrete for mapping to the same type, saves one from typing the template parameter
+    Opt<T> map(std::function<T(T t)> mapFunction) {
+      return Opt::map<T>(mapFunction);
+    };
+
+    // map<T,U>  - map from an Opt<T> to an Opt<U>
+    // if is None<T>, then returns None<U>
+    // else returns Some<U>
+    template<typename U>
+    Opt<U> map(std::function<U(T)> mapFunction) {
```

Comment:
Reiterating the comment we discussed where we could potentially use SFINAE to only add map to the Option type if T is copy constructible.

Reply:
Yeah this could be done. Like this?

```
    // map<T,U>  - map from an Opt<T> to an Opt<U>
    // if is None<T>, then returns None<U>
    // else returns Some<U>
    template<typename U , typename = std::enable_if_t<std::is_copy_constructible<T>::value == true>>
    Opt<U> map(std::function<U(T)> mapFunction) {
      static_assert(std::is_copy_constructible<U>::value == true, "Target type must be copy constructible.");
      if (m_pData) {
        auto u = mapFunction(*m_pData);
        U* pU = new U(u);
        return Opt<U>::Option(pU);
      }
      else {
        return Opt<U>::None();
      }
    };
```

Reply:
Precisely, except then you wouldn't need the static_assert but that is just nit picking ;).

Reply:
This is done. The static_assert is important because the other type (type U) also has to be copy constructible.